src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -426,25 +427,26 @@ private IQueryCache BuildQueryCache(string queryCacheName)
 					properties);
 		}
 
-		private ICacheConcurrencyStrategy GetCacheConcurrencyStrategy(
+		private Func<string, ICacheConcurrencyStrategy> GetCacheConcurrencyStrategy(
 			string cacheRegion,
 			string strategy,
 			bool isMutable,
-			Dictionary<Tuple<string, string>, ICacheConcurrencyStrategy> caches)
+			ConcurrentDictionary<Tuple<string, string>, ICacheConcurrencyStrategy> caches)
 		{
 			if (strategy == null || !settings.IsSecondLevelCacheEnabled)
 				return null;
 
-			var cacheKey = new Tuple<string, string>(cacheRegion, strategy);
-			if (caches.TryGetValue(cacheKey, out var cache)) 
-				return cache;
+			return tenantIdentifier =>
+			{
+				var cache = caches.GetOrAdd(
+					new Tuple<string, string>(tenantIdentifier + "|" + cacheRegion, strategy),
```

Comment:
Oups... I've pushed version with region name modification (and my current code is a mess so I won't update it now). But now `tenantIdentifier` is part of  cache key 
`new Tuple<string, string, string>(tenantIdentifier, cacheRegion, strategy)`